### PR TITLE
Start using the real Allegra and Mary eras

### PR DESCRIPTION
**Pull request by mrBliss**
Up until now, the Allegra and Mary eras part of `CardanoBlock` were simply
aliases of the Shelley era.

We now switch them over to the real Allegra and Mary eras. This means that some
things will be different for the eras, e.g., `TxBody`, `Script`, `Value`,
`UtxoPredicateFailure`, etc.

### CODE REVIEW

**Comment by mrBliss** [comment]
This is a draft because it depends on https://github.com/input-output-hk/ouroboros-network/pull/2679, which can't be merged until the generators in the ledger have been [generalised](https://github.com/input-output-hk/ouroboros-network/pull/2679#issuecomment-718704432). That's only blocking the tests, not the implementation.

@intricate @Jimbo4350 I have again left some TODOs for you marked with `TODO @node team`.

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs [hunk 1]
```diff
@@ -25,6 +25,8 @@ import           Control.Monad.Trans.Except.Extra (firstExceptT, handleIOExceptT
 import qualified Cardano.Binary as CBOR
 
 import qualified Shelley.Spec.Ledger.PParams as Shelley
+import           Shelley.Spec.Ledger.Scripts ()
+import           Cardano.Ledger.ShelleyMA.TxBody ()
```
Comment (mrBliss) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs:29]: I suppose that in the future the instances imported here will already be imported by other imports, e.g., `Cardano.Api.Typed`.

**Comment by mrBliss** [comment]
About the CI failure for 8.6.5: https://github.com/input-output-hk/cardano-ledger-specs/pull/1949#issuecomment-719511867

Note that hard forking will currently fail as the translations in the ledger have not yet been completed, e.g., https://github.com/input-output-hk/cardano-ledger-specs/pull/1951. Moreover, the second commit in https://github.com/input-output-hk/ouroboros-network/pull/2679 is also needed but not yet included. In other words: this draft PR just provides the *real* types, but not yet the ability to hard fork to Allegra or Mary.

File: cardano-node/src/Cardano/Tracing/OrphanInstances/Shelley.hs [hunk 2]
```diff
@@ -375,16 +389,84 @@ instance (ShelleyBasedEra era, Core.Value era ~ Coin)
              , "addrs"   .= addrs
              ]
 
+instance ToJSON (MA.Value era) where
+  toJSON = panic "TODO @node team: implement it. Should this be defined in cardano-ledger-specs?"
+
+-- TODO @node team: does this make sense? Note that the object can be empty.
+instance ToJSON MA.ValidityInterval where
```
Comment (Jimbo4350) [inline cardano-node/src/Cardano/Tracing/OrphanInstances/Shelley.hs:396]: We're currently using an intermediate type for this: https://github.com/input-output-hk/cardano-node/pull/2010/commits/f49a229a286baced9397d8a8cadf890a4c310a6d#diff-8595c8e945f3567404d3c90c4a43eaaf9dfb94aebecfac4810d39a20fc04b0ecR315

**Comment by mrBliss** [comment]
I rebased this against master of `cardano-ledger-specs` and the recently rebased https://github.com/input-output-hk/ouroboros-network/pull/2679

File: cardano-node/src/Cardano/Tracing/OrphanInstances/Shelley.hs [hunk 3]
```diff
@@ -389,16 +394,81 @@ instance (ShelleyBasedEra era, Core.Value era ~ Coin)
...
   | Set.null txIns = String "The transaction contains no inputs."
   | otherwise = String "The transaction contains inputs that do not exist in the UTxO set."
 
-renderValueNotConservedErr :: DeltaCoin -> DeltaCoin -> Value
-renderValueNotConservedErr consumed produced
-  | consumed > produced = String "This transaction has consumed more Lovelace than it has produced."
-  | consumed < produced = String "This transaction has produced more Lovelace than it has consumed."
-  | otherwise = String "Impossible: Somehow this error has occurred in spite of the transaction being balanced."
+-- TODO @node team: does this make sense?
+renderValueNotConservedErr :: Show val => val -> val -> Value
```
Comment (mrBliss) [inline cardano-node/src/Cardano/Tracing/OrphanInstances/Shelley.hs:469]: We use `renderValueNotConservedErr` with two instantiations of `val`:
* [Shelley and Allegra](https://github.com/input-output-hk/cardano-ledger-specs/blob/f1b1a0c3832aa3edffecebd744acd978d5d6611e/shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Utxo.hs#L137-L139): `Delta (Core.Value era)`, which reduces to `DeltaCoin`
* [Mary](https://github.com/input-output-hk/cardano-ledger-specs/blob/e1f839bff708e1938355b56ad3a7fe03930851ff/shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/Rules/Utxo.hs#L95-L97): `Core.Value era`, which reduces to [`Value`](https://github.com/input-output-hk/cardano-ledger-specs/blob/e1f839bff708e1938355b56ad3a7fe03930851ff/shelley-ma/impl/src/Cardano/Ledger/Mary/Value.hs#L75).

By specialising to the concrete types you could recover the more detailed error message.

**Comment by mrBliss** [comment]
https://github.com/input-output-hk/ouroboros-network/pull/2679 has been merged, so consensus is using the real era tags. I have rebased this PR and it is now ready for review (it still contains a few holes/questions for the node team!).

**Review by intricate — APPROVED** [review]

**Review by Jimbo4350 — APPROVED** [review]

**Pull request merged**